Comment on a line of a diff

bobsilverberg opened the pull request:
Fixes #989 

This allows for commenting on a normal or inserted line, but not a deleted line as there are issues that need to be ironed out on the server side to enable that. An "add comment" button will appear when hovering over the line number, and any comments added will appear underneath the line, just as they do in the browse view.

Screenshot:

![2019-11-21 16 35 46](https://user-images.githubusercontent.com/142755/69378824-2d610600-0c7d-11ea-9ce6-1b6a0dc674f7.gif)

bobsilverberg commented:
@kumar303 Can you please take a look at this and let me know what you think, both of the code and the UX?

bobsilverberg commented on src/components/DiffView/styles.module.scss:
```diff
@@ -44,6 +44,10 @@
     background-color: $light-gray;
   }
+
+  .diff-gutter:hover button.btn {
```
I had to do this hacky thing in order to target the button that is inside the gutter. The button is defined in a different file. Maybe there's a better way of doing this?

kumar303 replied:
I think you just needed to wrap the whole thing (including the line number) in `Commentable`. This is working for me:

<details>
<summary>Patch</summary>

```diff
diff --git a/src/components/DiffView/index.tsx b/src/components/DiffView/index.tsx
index 3922d61..f7db532 100644
--- a/src/components/DiffView/index.tsx
+++ b/src/components/DiffView/index.tsx
@@ -295,22 +295,25 @@ export class DiffViewBase extends React.Component<Props> {
     const { isInsert, lineNumber } = change;
     const changeKey = getChangeKey(change);
 
-    let gutter = wrapInAnchor(renderDefault());
+    const defaultGutter = wrapInAnchor(renderDefault());
+    let gutter = defaultGutter;
     if (enableCommenting && isInsert && lineNumber && side === 'new') {
       gutter = (
-        <div className={styles.gutter}>
-          {gutter}
-          <Commentable
-            as="span"
-            id={changeKey}
-            className={styles.commentButton}
-            line={lineNumber}
-            fileName={version.selectedPath}
-            versionId={version.id}
-          >
-            {(addCommentButton) => <>{addCommentButton}</>}
-          </Commentable>
-        </div>
+        <Commentable
+          as="div"
+          id={changeKey}
+          className={styles.gutter}
+          line={lineNumber}
+          fileName={version.selectedPath}
+          versionId={version.id}
+        >
+          {(addCommentButton) => (
+            <>
+              {defaultGutter}
+              <span className={styles.commentButton}>{addCommentButton}</span>
+            </>
+          )}
+        </Commentable>
       );
     }
 
diff --git a/src/components/DiffView/styles.module.scss b/src/components/DiffView/styles.module.scss
index bae2ff5..04a8900 100644
--- a/src/components/DiffView/styles.module.scss
+++ b/src/components/DiffView/styles.module.scss
@@ -44,10 +44,6 @@
   .diff-line .diff-code-selected {
     background-color: $light-gray;
   }
-
-  .diff-gutter:hover button.btn {
-    visibility: visible;
-  }
 }
 
 .extraMessages {
```

</details>

Ideally, we would wrap as much as possible in `Commentable` but I don't know if it's possible to do much more.

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -265,6 +284,39 @@ export class DiffViewBase extends React.Component<Props> {
+      gutter = (
+        <div className={styles.gutter}>
+          {gutter}
+          <Commentable
```
This approach seems to work but it would be nicer if one could hover over a line (not just the gutter) and see the comment button appear. This would convey that it's possible to add comments since that may not be obvious. 

To do that, we'd need to replace [the whole `<tr>`](https://github.com/otakustay/react-diff-view/blob/7a97efa2b1ed11c79cd755ba2fb0475de82b09a7/src/Hunk/UnifiedHunk/UnifiedChange.js#L80) with `<Commentable as="tr" />`. I don't think that's possible but it might be worth filing a feature request for it. They are super fast about fixing things.

But, yeah, the `renderGutter()` approach still works and I think it's worth pursuing that as a quick fix.

bobsilverberg replied:
Good suggestion. I opened https://github.com/otakustay/react-diff-view/issues/70.

kumar303 replied:
@bobsilverberg thanks for opening an issue. Rather than "wrap" the `<tr>` we really want to provider our own `<tr>` tag with custom props, like a class name. Maybe that's obvious enough in your issue description but I wanted to point it out since it's not possible to wrap `<tr>` tags, it will break the table layout.

bobsilverberg replied:
He's given me some suggestions of how to make this work. I think I'm going to try to land it first the way it is, with the button in the gutter, as we're still going to need that even if I also use `renderToken` to allow for hovering over the code area. I will file a follow-up to look into adding `renderToken` as well.

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,29 @@ export class DiffViewBase extends React.Component<Props> {
+            <CommentList
+              addonId={version.addon.id}
+              fileName={version.selectedPath}
+              line={line || null}
```
I guess TS wanted you to fall back to `null`? This should not be set to `null` so maybe you need to throw an error instead.

bobsilverberg replied:
I missed this one when I was adjusting the one below. I only have to gate this on `line` and all is fine.

bobsilverberg commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,29 @@ export class DiffViewBase extends React.Component<Props> {
+      );
       if (messages && messages.length) {
-        widget = (
+        widget = widget && (
```
I thought what I did worked, but I've changed it to join them together in a react fragment as per your comment.

bobsilverberg commented:
Thanks for your help thus far, @kumar303. This is ready for another look. I'm not entirely sure of the approach I took for testing `renderGutter`, I had to do some weird things.

bobsilverberg commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,32 @@ export class DiffViewBase extends React.Component<Props> {
-      let widget = null;
+      let widget = (
+        <>
+          {enableCommenting && line && change.isInsert && (
```
I wasn't sure if we might end up with duplicates for normal lines, but I guess that's impossible? Duplicates would only be an insert + a delete?

kumar303 replied:
Hmm, I'm not sure. In any case, the normal view (when there's no "diff") should show comments.

bobsilverberg commented on src/components/DiffView/index.tsx:
```diff
@@ -265,6 +287,42 @@ export class DiffViewBase extends React.Component<Props> {
+
+    const defaultGutter = wrapInAnchor(renderDefault());
+    let gutter = defaultGutter;
+    if (enableCommenting && isInsert && lineNumber && side === 'new') {
```
It shows the form for me, with the change, so I'm not sure what the problem was, but it seems okay on my end. Thanks for the suggestion!

bobsilverberg commented on src/components/DiffView/styles.module.scss:
```diff
@@ -57,3 +57,12 @@
   padding: 1rem 1rem 0 1rem;
 }
+
+.gutter {
```
Nice, thanks!

bobsilverberg commented on src/components/DiffView/index.spec.tsx:
```diff
@@ -573,23 +578,44 @@ describe(__filename, () => {
     const widgets = renderAndGetWidgets({
-      diff: parseDiff(diffWithDeletions)[0],
+      diff,
+      enableCommenting: true,
```
I was trying to be explicit about what data is being used to turn it on, but yeah, I guess that's not really necessary if the default test state is to have comments enabled.

bobsilverberg commented on src/components/DiffView/index.spec.tsx:
```diff
@@ -627,12 +653,80 @@ describe(__filename, () => {
     );
   });
 
-  it('does not render widgets without linter messages', () => {
```
Not with the current implementation. It renders a widget with a `CommentList` for every line, even if there are no comments. This allows the `CommentList` to be the only thing that cares about whether there are comments or not, but I guess that could be non-performant. Do you think we should be checking in `DiffView` to see if there are comments? I haven't checked yet to see how much (if any) extra work that would be.

kumar303 replied:
I didn't do any performance testing but I think it should be fine 😬  This was by design to make it easy to lazy load comments (not saying it's a great idea, heh).

bobsilverberg commented on src/components/DiffView/index.spec.tsx:
```diff
@@ -910,4 +1004,141 @@ describe(__filename, () => {
+        if (shallowRender) {
+          return shallow(<div>{gutter}</div>);
+        }
+        return gutter;
```
I saw `simulateCommentable` when I was writing the tests, but TBH I was confused about what it did and how it worked. I'll have to take another look, but you're right, that simple change makes it all work!

bobsilverberg commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,32 @@ export class DiffViewBase extends React.Component<Props> {
 
-      let widget = null;
+      let widget = (
+        <>
```
I originally had it as:

```tsx
widget = widget && (
  <div className={styles.inlineLinterMessages}>
    {messages.map((msg) => {
      return <LinterMessage key={msg.uid} message={msg} inline />;
    })}
  </div>
);
```

but I recall that you asked it to be wrapped in a fragment. Maybe I misunderstood what you wanted. If neither the above is okay, nor what's in my current patch, what should the syntax be?

bobsilverberg replied:
Actually, that syntax above doesn't work. But I can't just remove the fragment. Is there a better way to do it?

kumar303 replied:
Why can't you just remove the fragment? I only see one component rendered inside the fragment. Removing it compiles OK:

```diff
diff --git a/src/components/DiffView/index.tsx b/src/components/DiffView/index.tsx
index 37b464a..d626909 100644
--- a/src/components/DiffView/index.tsx
+++ b/src/components/DiffView/index.tsx
@@ -204,20 +204,17 @@ export class DiffViewBase extends React.Component<Props> {
         messages = selectedMessageMap.byLine[line];
       }
 
-      let widget = (
-        <>
-          {enableCommenting && line && !change.isDelete && (
-            <CommentList
-              addonId={version.addon.id}
-              fileName={version.selectedPath}
-              line={line}
-              versionId={version.id}
-            >
-              {(commentList) => <div>{commentList}</div>}
-            </CommentList>
-          )}
-        </>
-      );
+      let widget =
+        enableCommenting && line && !change.isDelete ? (
+          <CommentList
+            addonId={version.addon.id}
+            fileName={version.selectedPath}
+            line={line}
+            versionId={version.id}
+          >
+            {(commentList) => <div>{commentList}</div>}
+          </CommentList>
+        ) : null;
       if (messages && messages.length) {
         widget = (
           <>

```

bobsilverberg replied:
Doh! I was looking at the *wrong* fragment. Although it was clear from the line number, for some reason I thought you were talking about the fragment on line 223, not line 208. My bad. I'll remove the one on 208.

bobsilverberg commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,32 @@ export class DiffViewBase extends React.Component<Props> {
+              line={line}
+              versionId={version.id}
+            >
+              {(commentList) => <div>{commentList}</div>}
```
Ok, I did this, and it works, but I find it pretty confusing to read. Maybe my test can be improved, or maybe we're stuck with the fact that we need to call `renderContent` using the `shell`, which just comes across as very cryptic in the test.

bobsilverberg commented:
Thanks for the review and all of the suggestions, @kumar303. I've addressed everything except for the two comments that remain unresolved for which I've added comments/questions. It's ready for another look.

bobsilverberg commented:
@kumar303 In case this has fallen off your radar, it's ready for another look.

kumar303 commented:
Yeah, it's in my queue, thanks. Could you also update the description when you get a chance?

kumar303 commented on src/components/DiffView/index.spec.tsx:
```diff
@@ -627,12 +656,105 @@ describe(__filename, () => {
+          commentList: <CommentListResult />,
+          root: widget,
+        });
+        const commentWidget = renderContent(shell.at(0));
```
Since there will only be one `<CommentList>` in the widget, you can simulate it with the default params:

```diff
diff --git a/src/components/DiffView/index.spec.tsx b/src/components/DiffView/index.spec.tsx
index 5ef8bf3..d5b93a6 100644
--- a/src/components/DiffView/index.spec.tsx
+++ b/src/components/DiffView/index.spec.tsx
@@ -683,11 +683,11 @@ describe(__filename, () => {
       for (const line of [firstLine, secondLine]) {
         const widget = renderWidget(diff.hunks, widgets, line);
 
-        const { shell, renderContent } = simulateCommentList({
+        const { renderContent } = simulateCommentList({
           commentList: <CommentListResult />,
           root: widget,
         });
-        const commentWidget = renderContent(shell.at(0));
+        const commentWidget = renderContent();
         expect(commentWidget.find(CommentListResult)).toHaveLength(1);
 
         const commentList = widget.find(CommentList);
```

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,29 @@ export class DiffViewBase extends React.Component<Props> {
+          line={line}
+          versionId={version.id}
+        >
+          {(commentList) => <div>{commentList}</div>}
```
This can be `{(allComments) => allComments}`. That is, there's no need for a div and in fact it's already [wrapped in one](https://github.com/mozilla/addons-code-manager/blob/83eb299a13464e5f25a1bbd47a4fae917eac3f6b/src/components/CommentList/index.tsx#L106). 

It may seem weird to need a render prop for this but if you consider wanting to render a container element around the comments but only when they exist, there's actually no other way to do that.

kumar303 commented on src/components/DiffView/index.spec.tsx:
```diff
@@ -910,4 +1032,146 @@ describe(__filename, () => {
+    }: Partial<RenderGutterParams> &
+      RenderParams & {
+        defaultGutter?: React.ReactElement;
+        shallowRender?: boolean;
```
This isn't used anymore

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -197,17 +204,29 @@ export class DiffViewBase extends React.Component<Props> {
       }
 
-      let widget = null;
+      let widget = enableCommenting && line && !change.isDelete && (
```
In the rare (and probably unlikely) case that line is `0` then the fragment below will put a zero in the HTML which is not what you want. I suggest using a ternary and setting it to `null` when false.

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -265,6 +284,42 @@ export class DiffViewBase extends React.Component<Props> {
+  }: RenderGutterParams) => {
+    const { enableCommenting, version } = this.props;
+
+    const { isDelete, lineNumber } = change;
```
This [part of the docs](https://github.com/otakustay/react-diff-view/#key-of-change) suggests maybe you should be using `oldLineNumber` for normal changes. Could you try to comment on some normal diffs and compare it in browse mode to make sure they appear on the right lines? It might be fine; I have a vague recollection that `lineNumber` and `oldLineNumber` are identical....maybe.

bobsilverberg replied:
I checked, and a comment added in Compare mode on a normal line does also appear in Browse mode for the same file and version, so I guess the current code is okay.

kumar303 commented on src/components/DiffView/index.tsx:
```diff
@@ -265,6 +284,42 @@ export class DiffViewBase extends React.Component<Props> {
+      gutter = (
+        <Commentable
+          as="div"
+          id={changeKey}
```
You don't have to define an ID here. It's only there in case the wrapped element (in this case `div`) needs an `id` attribute.
